vitrina/api/oauth.py

```diff
@@ -104,14 +113,107 @@ def retrieve_access_token_from_request(request: Request) -> str | None:
         return token_value
 
     @staticmethod
-    def retrieve_and_verify_token(request: Request) -> JWTClaims | None:
-        access_token = OAuthClientAuthenticator.retrieve_access_token_from_request(request)
-        if not access_token:
+    def download_jwk_verification_keys() -> JWK_KEY | dict[str, JWK_KEY]:
+        if not settings.OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_PATH:
+            logger.error(
+                "Cannot download public JWK verification keys, OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_PATH is not set."
+            )
+            return {}
+        response = requests.get(settings.OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_URL)
+        if not response.ok:
+            logger.error(
+                f"Failed to download public JWK verification keys from {settings.OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_URL=} - {response.status_code=} {response.text=}"
+            )
+            return {}
+        return response.json()
+
+    @classmethod
+    def get_jwk_verification_keys(cls, check_cache: bool = True) -> list[Key]:
+        if settings.OAUTH_SERVER_PUBLIC_JWK_JSON:
+            verification_keys = settings.OAUTH_SERVER_PUBLIC_JWK_JSON.get("keys") or [
+                settings.OAUTH_SERVER_PUBLIC_JWK_JSON
+            ]
+            return [JsonWebKey.import_key(verification_key) for verification_key in verification_keys]
+        elif settings.OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_PATH:
+            if check_cache and (cached_verification_keys := cache.get(cls.JWK_VERIFICATION_KEYS_CACHE_KEY)):
+                return cached_verification_keys
+            verification_keys = cls.download_jwk_verification_keys()
+            if verification_keys:
+                verification_keys = verification_keys.get("keys") or [verification_keys]
+                verification_keys = [JsonWebKey.import_key(verification_key) for verification_key in verification_keys]
+                cache.set(
+                    cls.JWK_VERIFICATION_KEYS_CACHE_KEY,
+                    verification_keys,
+                    timeout=settings.OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_CACHE_TIMEOUT,
+                )
+                cache.set(cls.BACKUP_JWK_VERIFICATION_KEYS_CACHE_KEY, verification_keys)
+                return verification_keys
+            else:
+                logger.warning("Could not retrieve public JWK verification keys, using backup cached ones.")
+                return cache.get(cls.BACKUP_JWK_VERIFICATION_KEYS_CACHE_KEY) or []
+        raise ImproperlyConfigured(
+            "No public JWK verification keys configured (OAUTH_SERVER_PUBLIC_JWK_JSON/OAUTH_SERVER_PUBLIC_JWK_DOWNLOAD_PATH)."
+        )
+
+    @staticmethod
+    def decode_unverified_header(token: str) -> dict[str, Any]:
+        try:
+            if isinstance(token, bytes):
```

Comment:
The typehint seems to be amiss:


```suggestion
    def decode_unverified_header(token: str | bytes) -> dict[str, Any]:
        try:
            if isinstance(token, bytes):  # Based on this check
```